In [ ]:
import warnings
warnings.filterwarnings("ignore")

import numpy as np
import pandas as pd
from sklearn.model_selection import train_test_split, cross_val_score
from sklearn.preprocessing import StandardScaler
from sklearn.ensemble import StackingRegressor
from sklearn.linear_model import RidgeCV
from sklearn.metrics import r2_score, root_mean_squared_error
import optuna

# GPU-compatible libraries
import xgboost as xgb
import lightgbm as lgb
from catboost import CatBoostRegressor

# ======================
# 1) Dati + preprocessing
# ======================
from sklearn.datasets import fetch_california_housing
raw = fetch_california_housing()
data = pd.DataFrame(raw.data, columns=raw.feature_names)
data["MedHouseVal"] = raw.target

# Feature engineering
data["Ave_Room_Bed"] = data["AveRooms"] / (data["AveBedrms"] + 1e-6)
data["RoomsPerPerson"] = data["AveRooms"] / (data["Population"] + 1.0)
data["Lat_Long"] = data["Latitude"] * data["Longitude"]

X = data.drop(columns=["MedHouseVal"])
y = data["MedHouseVal"].values

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)
scaler = StandardScaler()
X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

# ======================
# 2) Funzioni obiettivo Optuna (GPU)
# ======================
def objective_xgb(trial):
    param = {
        "n_estimators": trial.suggest_int("n_estimators", 100, 500),
        "max_depth": trial.suggest_int("max_depth", 3, 10),
        "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.2, log=True),
        "subsample": trial.suggest_float("subsample", 0.6, 1.0),
        "colsample_bytree": trial.suggest_float("colsample_bytree", 0.6, 1.0),
        "reg_alpha": trial.suggest_float("reg_alpha", 1e-6, 10.0, log=True),
        "reg_lambda": trial.suggest_float("reg_lambda", 1e-6, 10.0, log=True),
        "random_state": 42,
        "tree_method": "gpu_hist",      # GPU
        "predictor": "gpu_predictor",   # GPU
        "n_jobs": -1
    }
    model = xgb.XGBRegressor(**param)
    scores = cross_val_score(model, X_train_scaled, y_train, cv=3,
                             scoring="neg_root_mean_squared_error", n_jobs=1)
    return -scores.mean()

def objective_lgb(trial):
    param = {
        "n_estimators": trial.suggest_int("n_estimators", 100, 1000),
        "num_leaves": trial.suggest_int("num_leaves", 16, 256),
        "max_depth": trial.suggest_int("max_depth", 3, 16),
        "learning_rate": trial.suggest_float("learning_rate", 0.005, 0.15, log=True),
        "subsample": trial.suggest_float("subsample", 0.6, 1.0),
        "colsample_bytree": trial.suggest_float("colsample_bytree", 0.6, 1.0),
        "reg_alpha": trial.suggest_float("reg_alpha", 1e-6, 10.0, log=True),
        "reg_lambda": trial.suggest_float("reg_lambda", 1e-6, 10.0, log=True),
        "random_state": 42,
        "device": "gpu",   # GPU
        "n_jobs": -1
    }
    model = lgb.LGBMRegressor(**param)
    scores = cross_val_score(model, X_train_scaled, y_train, cv=3,
                             scoring="neg_root_mean_squared_error", n_jobs=1)
    return -scores.mean()

def objective_cat(trial):
    param = {
        "iterations": trial.suggest_int("iterations", 500, 1500),
        "learning_rate": trial.suggest_float("learning_rate", 0.01, 0.1),
        "depth": trial.suggest_int("depth", 4, 10),
        "l2_leaf_reg": trial.suggest_float("l2_leaf_reg", 1, 10),
        "random_state": 42,
        "task_type": "GPU",   # GPU
        "devices": "0:1",     # GPU device(s)
        "verbose": 0
    }
    model = CatBoostRegressor(**param)
    scores = cross_val_score(model, X_train_scaled, y_train, cv=3,
                             scoring="neg_root_mean_squared_error", n_jobs=1)
    return -scores.mean()

# ======================
# 3) Esegui ottimizzazione Optuna
# ======================
study_xgb = optuna.create_study(direction="minimize")
study_xgb.optimize(objective_xgb, n_trials=20, show_progress_bar=False)

study_lgb = optuna.create_study(direction="minimize")
study_lgb.optimize(objective_lgb, n_trials=20, show_progress_bar=False)

study_cat = optuna.create_study(direction="minimize")
study_cat.optimize(objective_cat, n_trials=15, show_progress_bar=False)

best_xgb_params = study_xgb.best_params
best_lgb_params = study_lgb.best_params
best_cat_params = study_cat.best_params

print("Best XGB:", best_xgb_params)
print("Best LGB:", best_lgb_params)
print("Best Cat:", best_cat_params)

# ======================
# 4) Costruisci modelli finali
# ======================
xgb_est = xgb.XGBRegressor(**best_xgb_params, random_state=42, tree_method="gpu_hist", predictor="gpu_predictor")
lgb_est = lgb.LGBMRegressor(**best_lgb_params, random_state=42, device="gpu")
cat_est = CatBoostRegressor(**best_cat_params, task_type="GPU", devices="0:1", verbose=0)

xgb_est.fit(X_train_scaled, y_train)
lgb_est.fit(X_train_scaled, y_train)
cat_est.fit(X_train_scaled, y_train)

# ======================
# 5) Stacking
# ======================
estimators = [("xgb", xgb_est), ("lgb", lgb_est), ("cat", cat_est)]
stack = StackingRegressor(
    estimators=estimators,
    final_estimator=RidgeCV(),
    cv=5,
    n_jobs=-1
)

stack.fit(X_train_scaled, y_train)
pred_stack = stack.predict(X_test_scaled)
rmse_stack = root_mean_squared_error(y_test, pred_stack)
r2_stack = r2_score(y_test, pred_stack)

print(f"STACK RMSE: {rmse_stack:.4f}, R2: {r2_stack:.4f}")


[I 2025-12-09 17:21:25,495] A new study created in memory with name: no-name-6b5a5910-3b7c-4cf8-a542-e7a322688344
[I 2025-12-09 17:21:27,916] Trial 0 finished with value: 0.49544998943977525 and parameters: {'n_estimators': 298, 'max_depth': 5, 'learning_rate': 0.023017410539018155, 'subsample': 0.8837345337915046, 'colsample_bytree': 0.8253222242186982, 'reg_alpha': 1.5004425456864625e-06, 'reg_lambda': 8.342951058259236e-05}. Best is trial 0 with value: 0.49544998943977525.
[I 2025-12-09 17:21:32,479] Trial 1 finished with value: 0.48221668798081163 and parameters: {'n_estimators': 443, 'max_depth': 6, 'learning_rate': 0.014470181162732272, 'subsample': 0.6536910220959432, 'colsample_bytree': 0.8934231330195344, 'reg_alpha': 0.002938285614431645, 'reg_lambda': 4.578121933675406e-06}. Best is trial 1 with value: 0.48221668798081163.
[I 2025-12-09 17:21:44,610] Trial 2 finished with value: 0.4670621527614512 and parameters: {'n_estimators': 283, 'max_depth': 10, 'learning_rate': 0.0252

[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000902 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8


[I 2025-12-09 17:23:17,822] Trial 0 finished with value: 0.45896222015991595 and parameters: {'n_estimators': 420, 'num_leaves': 43, 'max_depth': 16, 'learning_rate': 0.1458620329805018, 'subsample': 0.9860192568465089, 'colsample_bytree': 0.9249903968389634, 'reg_alpha': 0.4948812706321945, 'reg_lambda': 0.04163744219389}. Best is trial 0 with value: 0.45896222015991595.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000852 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8


[I 2025-12-09 17:23:31,773] Trial 1 finished with value: 0.5156244821371937 and parameters: {'n_estimators': 714, 'num_leaves': 21, 'max_depth': 7, 'learning_rate': 0.005521751811186665, 'subsample': 0.694556410623593, 'colsample_bytree': 0.6261816266200942, 'reg_alpha': 0.09575273435280486, 'reg_lambda': 4.463656451382153}. Best is trial 0 with value: 0.45896222015991595.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000649 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:23:36,939] Trial 2 finished with value: 0.5095462089628945 and parameters: {'n_estimators': 835, 'num_leaves': 126, 'max_depth': 3, 'learning_rate': 0.018396880678429466, 'subsample': 0.6780447583224005, 'colsample_bytree': 0.976755559230563, 'reg_alpha': 0.001350983803366993, 'reg_lambda': 2.0034832436942667}. Best is trial 0 with value: 0.45896222015991595.


[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No f

[I 2025-12-09 17:23:44,678] Trial 3 finished with value: 0.5233449947673047 and parameters: {'n_estimators': 345, 'num_leaves': 86, 'max_depth': 5, 'learning_rate': 0.011670198661767073, 'subsample': 0.6446628831465313, 'colsample_bytree': 0.7504278029881482, 'reg_alpha': 0.4901688531924287, 'reg_lambda': 1.5713329881388758}. Best is trial 0 with value: 0.45896222015991595.


[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histog

[I 2025-12-09 17:24:49,080] Trial 4 finished with value: 0.46305968111890344 and parameters: {'n_estimators': 985, 'num_leaves': 93, 'max_depth': 11, 'learning_rate': 0.14169095864486408, 'subsample': 0.6168347118476316, 'colsample_bytree': 0.800064622692129, 'reg_alpha': 3.3858682009589943e-05, 'reg_lambda': 0.18961948916952523}. Best is trial 0 with value: 0.45896222015991595.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000671 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:26:14,190] Trial 5 finished with value: 0.4462277401601431 and parameters: {'n_estimators': 696, 'num_leaves': 146, 'max_depth': 16, 'learning_rate': 0.015701868558545173, 'subsample': 0.9821995708193652, 'colsample_bytree': 0.7329112450288671, 'reg_alpha': 2.9210923584938802e-06, 'reg_lambda': 0.004166223880604601}. Best is trial 5 with value: 0.4462277401601431.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.001399 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:27:48,825] Trial 6 finished with value: 0.4517825270414048 and parameters: {'n_estimators': 974, 'num_leaves': 197, 'max_depth': 13, 'learning_rate': 0.029588050834028223, 'subsample': 0.8121458732836352, 'colsample_bytree': 0.787158310513935, 'reg_alpha': 2.686103355661728e-05, 'reg_lambda': 0.10353969608284662}. Best is trial 5 with value: 0.4462277401601431.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000955 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:28:13,924] Trial 7 finished with value: 0.468123521889299 and parameters: {'n_estimators': 856, 'num_leaves': 177, 'max_depth': 6, 'learning_rate': 0.015437720816877451, 'subsample': 0.8048816153691726, 'colsample_bytree': 0.9023672634731277, 'reg_alpha': 0.0005592411267785228, 'reg_lambda': 0.010169764754870336}. Best is trial 5 with value: 0.4462277401601431.


[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transf

[I 2025-12-09 17:28:35,392] Trial 8 finished with value: 0.46091666317129265 and parameters: {'n_estimators': 306, 'num_leaves': 100, 'max_depth': 12, 'learning_rate': 0.08708483250015508, 'subsample': 0.9400960661232434, 'colsample_bytree': 0.9899314153921717, 'reg_alpha': 0.0004384031675436601, 'reg_lambda': 0.000690912498028453}. Best is trial 5 with value: 0.4462277401601431.


[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000885 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Usi

[I 2025-12-09 17:28:55,242] Trial 9 finished with value: 0.5784098654782432 and parameters: {'n_estimators': 230, 'num_leaves': 90, 'max_depth': 16, 'learning_rate': 0.007156116373706425, 'subsample': 0.7271643494703691, 'colsample_bytree': 0.8832929417900461, 'reg_alpha': 6.226680139416757e-05, 'reg_lambda': 4.716021422828602}. Best is trial 5 with value: 0.4462277401601431.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000914 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:29:31,141] Trial 10 finished with value: 0.44949868895673123 and parameters: {'n_estimators': 560, 'num_leaves': 256, 'max_depth': 9, 'learning_rate': 0.03881678669224719, 'subsample': 0.8990264535916935, 'colsample_bytree': 0.6474112794386593, 'reg_alpha': 1.3204537304974887e-06, 'reg_lambda': 9.895518843453215e-06}. Best is trial 5 with value: 0.4462277401601431.


[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000764 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:30:04,070] Trial 11 finished with value: 0.44937795266890346 and parameters: {'n_estimators': 575, 'num_leaves': 242, 'max_depth': 8, 'learning_rate': 0.031348823220764405, 'subsample': 0.8947727632654904, 'colsample_bytree': 0.6339072722477577, 'reg_alpha': 1.9574951529973835e-06, 'reg_lambda': 5.1375046641556e-06}. Best is trial 5 with value: 0.4462277401601431.


[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000950 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:30:39,940] Trial 12 finished with value: 0.4501016866688403 and parameters: {'n_estimators': 582, 'num_leaves': 246, 'max_depth': 9, 'learning_rate': 0.05089586954673087, 'subsample': 0.9084043458417898, 'colsample_bytree': 0.708216433636893, 'reg_alpha': 1.2463992988156254e-06, 'reg_lambda': 0.00021147532797823436}. Best is trial 5 with value: 0.4462277401601431.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000931 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:32:03,191] Trial 13 finished with value: 0.4441239790854971 and parameters: {'n_estimators': 682, 'num_leaves': 192, 'max_depth': 14, 'learning_rate': 0.021140088018825578, 'subsample': 0.8606216739964195, 'colsample_bytree': 0.6728165771498259, 'reg_alpha': 0.020881193973055753, 'reg_lambda': 1.8991299179422203e-06}. Best is trial 13 with value: 0.4441239790854971.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000839 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:33:51,224] Trial 14 finished with value: 0.44897161048769413 and parameters: {'n_estimators': 718, 'num_leaves': 174, 'max_depth': 14, 'learning_rate': 0.009111903619300487, 'subsample': 0.9950069402806191, 'colsample_bytree': 0.7037050111887921, 'reg_alpha': 0.023571296196906438, 'reg_lambda': 4.8183912975441764e-05}. Best is trial 13 with value: 0.4441239790854971.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000914 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:35:00,173] Trial 15 finished with value: 0.4540377857282289 and parameters: {'n_estimators': 698, 'num_leaves': 205, 'max_depth': 14, 'learning_rate': 0.019504096800534134, 'subsample': 0.8471566058982276, 'colsample_bytree': 0.6994424580019375, 'reg_alpha': 8.548467492740437, 'reg_lambda': 0.0018285158484022575}. Best is trial 13 with value: 0.4441239790854971.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000933 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:35:52,225] Trial 16 finished with value: 0.4517432578746381 and parameters: {'n_estimators': 467, 'num_leaves': 147, 'max_depth': 15, 'learning_rate': 0.057194773092008426, 'subsample': 0.7483362020835878, 'colsample_bytree': 0.8286092512727443, 'reg_alpha': 0.011415647969480436, 'reg_lambda': 1.6942082611168842e-06}. Best is trial 13 with value: 0.4441239790854971.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000683 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:37:19,106] Trial 17 finished with value: 0.4459353265578591 and parameters: {'n_estimators': 831, 'num_leaves': 143, 'max_depth': 11, 'learning_rate': 0.0122347853391202, 'subsample': 0.8562018450584647, 'colsample_bytree': 0.601419794296681, 'reg_alpha': 0.008119776180804292, 'reg_lambda': 0.00017379295918360101}. Best is trial 13 with value: 0.4441239790854971.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.000742 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8


[I 2025-12-09 17:37:38,851] Trial 18 finished with value: 0.6322474378787681 and parameters: {'n_estimators': 100, 'num_leaves': 216, 'max_depth': 11, 'learning_rate': 0.011412354685182367, 'subsample': 0.8516700725974515, 'colsample_bytree': 0.6666084387148014, 'reg_alpha': 0.006215137655648868, 'reg_lambda': 8.510800058520792e-05}. Best is trial 13 with value: 0.4441239790854971.


[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2602
[LightGBM] [Info] Number of data points in the train set: 11008, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram bin entry: 8
[LightGBM] [Info] 11 dense feature groups (0.13 MB) transferred to GPU in 0.001020 secs. 0 sparse feature groups
[LightGBM] [Info] Start training from score 2.064393
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits wit

[I 2025-12-09 17:39:06,687] Trial 19 finished with value: 0.44332805898105887 and parameters: {'n_estimators': 835, 'num_leaves': 172, 'max_depth': 12, 'learning_rate': 0.022939954463489023, 'subsample': 0.7602055671751626, 'colsample_bytree': 0.6696394664111355, 'reg_alpha': 0.07246715421257888, 'reg_lambda': 1.5988379921499036e-05}. Best is trial 19 with value: 0.44332805898105887.
[I 2025-12-09 17:39:06,688] A new study created in memory with name: no-name-876fd978-53a7-492a-ac34-b6503008d406
[I 2025-12-09 17:40:04,096] Trial 0 finished with value: 0.4532930393539933 and parameters: {'iterations': 1058, 'learning_rate': 0.09350716513068133, 'depth': 7, 'l2_leaf_reg': 4.925167659419896}. Best is trial 0 with value: 0.4532930393539933.
[I 2025-12-09 17:40:50,369] Trial 1 finished with value: 0.4605364654195936 and parameters: {'iterations': 1165, 'learning_rate': 0.081932141015925, 'depth': 5, 'l2_leaf_reg': 9.17064849895506}. Best is trial 0 with value: 0.4532930393539933.
[I 2025-12

Best XGB: {'n_estimators': 360, 'max_depth': 9, 'learning_rate': 0.03821528290773023, 'subsample': 0.683533094403276, 'colsample_bytree': 0.7187917338264955, 'reg_alpha': 3.1906582182306593, 'reg_lambda': 0.022554881280094124}
Best LGB: {'n_estimators': 835, 'num_leaves': 172, 'max_depth': 12, 'learning_rate': 0.022939954463489023, 'subsample': 0.7602055671751626, 'colsample_bytree': 0.6696394664111355, 'reg_alpha': 0.07246715421257888, 'reg_lambda': 1.5988379921499036e-05}
Best Cat: {'iterations': 1470, 'learning_rate': 0.06684189434508243, 'depth': 10, 'l2_leaf_reg': 6.978097064401743}
[LightGBM] [Info] This is the GPU trainer!!
[LightGBM] [Info] Total Bins 2603
[LightGBM] [Info] Number of data points in the train set: 16512, number of used features: 11
[LightGBM] [Info] Using GPU Device: NVIDIA GeForce RTX 4060 Ti, Vendor: NVIDIA Corporation
[LightGBM] [Info] Compiling OpenCL Kernel with 256 bins...
[LightGBM] [Info] GPU programs have been built
[LightGBM] [Info] Size of histogram b

Exception ignored on calling ctypes callback function: <function _log_callback at 0x000002777F66CCC0>
Traceback (most recent call last):
  File "c:\Python312\Lib\site-packages\lightgbm\basic.py", line 287, in _log_callback
    def _log_callback(msg: bytes) -> None:
    
KeyboardInterrupt: 


No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with positive gain, best gain: -inf
[LightGBM] [Warning] No further splits with po

**RISULTATO**

RMSE Baseline:    0.4718
RMSE Ottimizzato: 0.4300

param = {
        'verbosity': 0,
        'objective': 'reg:squarederror',
        'eval_metric': 'rmse', # Metrica per early stopping e pruning
        'tree_method': 'hist',
        
        # --- Struttura ---
        # California ha interazioni complesse, permettiamo alberi più profondi
        'max_depth': trial.suggest_int('max_depth', 2, 15),
        # Importante per evitare overfitting su outlier di prezzo
        'min_child_weight': trial.suggest_int('min_child_weight', 1, 20),
        'gamma': trial.suggest_float('gamma', 1e-8, 1.0, log=True),
        
        # --- Randomness ---
        'subsample': trial.suggest_float('subsample', 0.6, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.6, 1.0),
        
        # --- Regolarizzazione ---
        # Fondamentale nella regressione per non inseguire i prezzi estremi
        'lambda': trial.suggest_float('lambda', 1e-9, 25.0, log=True),
        'alpha': trial.suggest_float('alpha', 1e-9, 25.0, log=True),
        
        # --- Learning ---
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.3, log=True),
    }

study.optimize(objective, n_trials=70, timeout=600)

=============================================================================== <br>
RMSE Baseline:    0.4718
RMSE Ottimizzato: 0.4310

con scaler

param = {
        'verbosity': 0,
        'objective': 'reg:squarederror',
        'eval_metric': 'rmse', # Metrica per early stopping e pruning
        'tree_method': 'hist',
        
        # --- Struttura ---
        # California ha interazioni complesse, permettiamo alberi più profondi
        'max_depth': trial.suggest_int('max_depth', 2, 20),
        # Importante per evitare overfitting su outlier di prezzo
        'min_child_weight': trial.suggest_int('min_child_weight', 1, 25),
        'gamma': trial.suggest_float('gamma', 1e-9, 1.5, log=True),
        
        # --- Randomness ---
        'subsample': trial.suggest_float('subsample', 0.6, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.6, 1.0),
        
        # --- Regolarizzazione ---
        # Fondamentale nella regressione per non inseguire i prezzi estremi
        'lambda': trial.suggest_float('lambda', 1e-6, 20.0, log=True),
        'alpha': trial.suggest_float('alpha', 1e-6, 20.0, log=True),
        
        # --- Learning ---
        'learning_rate': trial.suggest_float('learning_rate', 0.01, 0.8),
    }

study.optimize(objective, n_trials=200, timeout=600)

=============================================================================== <br>

RMSE Baseline:    0.3522
RMSE Ottimizzato: 0.3379

con scaler

study.optimize(objective, n_trials=200, timeout=600)

param = {
        'verbosity': 0,
        'objective': 'reg:squarederror',
        'eval_metric': 'rmse', # Metrica per early stopping e pruning
        'tree_method': 'hist',
        
        # --- Struttura ---
        # California ha interazioni complesse, permettiamo alberi più profondi
        'max_depth': trial.suggest_int('max_depth', 2, 20),
        # Importante per evitare overfitting su outlier di prezzo
        'min_child_weight': trial.suggest_int('min_child_weight', 1, 25),
        'gamma': trial.suggest_float('gamma', 1e-7, 1.5, log=True),
        
        # --- Randomness ---
        'subsample': trial.suggest_float('subsample', 0.6, 1.0),
        'colsample_bytree': trial.suggest_float('colsample_bytree', 0.6, 1.0),
        
        # --- Regolarizzazione ---
        # Fondamentale nella regressione per non inseguire i prezzi estremi
        'lambda': trial.suggest_float('lambda', 1e-9, 15.0, log=True),
        'alpha': trial.suggest_float('alpha', 1e-9, 15.0, log=True),
        
        # --- Learning ---
        'learning_rate': trial.suggest_float('learning_rate', 0.001, 0.6),
    }

#rimozione outlier
numeric_cols = data.select_dtypes(include=["float", "int"]).columns.tolist()

for col in numeric_cols:
    Q1 = data[col].quantile(0.3)
    Q3 = data[col].quantile(0.7)
    IQR = Q3 - Q1
    lower, upper = Q1 - 1.5 * IQR, Q3 + 1.5 * IQR

    data[col] = np.where(data[col].between(lower, upper), data[col], np.nan)

data = data.dropna()

#features combinate
data['Ave_Room_Bed'] = data['AveRooms'] / data['AveBedrms']
data['HouseRooms'] = data['HouseAge'] / data['AveRooms']

Miglior RMSE trovato da Optuna (Validation): 0.3544
Migliori parametri: {'max_depth': 13, 'min_child_weight': 23, 'gamma': 3.8838420870962745e-06, 'subsample': 0.7570628531701007, 'colsample_bytree': 0.9109736816364161, 'lambda': 0.02307771384295654, 'alpha': 0.9454327224813569, 'learning_rate': 0.06885420667941514}